### PR TITLE
fix(animations): compensate for layout changes and synchronize tween phases

### PR DESCRIPTION
**Pull request by bao-ho**
**Checklist**

- [x] tests added
- [x] commits conform to the [commit guidelines](./CONTRIBUTING.md#commit)
- [ ] documentation updated (not relevant)

## Support animations
- Add properties to nodes of some components so that the `trackBy` functions can grab those properties as IDs during tweening
  - Axes' tick marks and labels: add tick values
  - Grid lines: add tick values and grid lines' direction (x or y)
- Component-factory.js
  - The issue: as soon as tweening begins, the layout changes immediately to a new layout while the displaying nodes are still at the same relative positions to the old layout. This makes the nodes "jump" at the beginning of the animations.
    - To fix this, we can compesate for the layout changes by adjusting the relative positions of the displaying nodes. For example, if the new layout of the point component (the central area) jumps 10px to the left compared to the old layout, we shift the points 10px to the right, making the absolute positions of the points stay the same.
    - We call the `compensateForLayoutChanges` function of each component before doing tweening
    - A new object, `preComputedRect` is introduced to save the old layout. The new layout is in `instanceContext.rect.computed`
- Tween.js:
  - Always run all three tween phases (exit, update, and enter) even though they may be empty. This is to synchronize animations of different components, avoiding for example updating phase of one component running during exiting phase of another component.
- Unit tests: ✅

### Verification (sn scatter plot)

https://user-images.githubusercontent.com/70384379/140223607-3b27f726-beb5-4e97-85a3-aa49c9ac3f98.mov

### CODE REVIEW

File: packages/picasso.js/src/core/component/component-factory.js [hunk 1]
```diff
@@ -489,6 +491,7 @@ function componentFactory(definition, context = {}) {
     });
 
     if (currentNodes && settings.animations && settings.animations.enabled) {
+      settings.animations.compensateForLayoutChanges(currentNodes, instanceContext.rect.computed, preComputedRect);
```
Comment (veinfors) [inline packages/picasso.js/src/core/component/component-factory.js:494]: First of all - don't we need to check if the `compensateForLayoutChanges` function exist before we call it here?

Then, to make it a little easier to work with, we rename the input parameters and send them in as an object, something like this:
```suggestion
      settings.animations.compensateForLayoutChanges({currentNodes, currentRect: instanceContext.rect.computed, newRect: preComputedRect});
```

File: packages/picasso.js/src/core/component/tween.js [hunk 2]
```diff
@@ -58,26 +58,26 @@ function tween({ old, current }, { renderer }, config) {
           exited.ips.push(interpolateObject(ids[key], extend({}, ids[key], { r: 0.0001, opacity: 0 })));
         }
       });
-      if (exited.ips.length) {
+      if (exited.ips.length >= 0) {
```
Comment (veinfors) [inline packages/picasso.js/src/core/component/tween.js:61]: Is it necessary to check like this? Why if so?

Comment (bao-ho) [reply packages/picasso.js/src/core/component/tween.js:61]: yeah, it should be written better.

File: packages/picasso.js/src/core/component/component-factory.js (hunk 1 above)
Comment (bao-ho) [reply packages/picasso.js/src/core/component/component-factory.js:494]: good points!

**Review by veinfors — APPROVED** [review]

**Pull request merged**